File: sdk/kotlin/src/main/kotlin/com/kycvault/DIDResolver.kt
package com.kycvault

class DIDResolver(private val client: KYCVaultClient) {
    companion object {
        private val SUPPORTED_METHODS = listOf("key", "web", "ethr", "ion", "indy")

        fun parseDID(did: String): Pair<String, String> {
            val parts = did.split(":")
            require(parts.size >= 3 && parts[0] == "did") { "Invalid DID: $did" }
            return Pair(parts[1], parts.drop(2).joinToString(":"))
        }

        fun isValidDID(did: String): Boolean = try {
            parseDID(did)
            true
        } catch (_: Exception) {
            false
        }

        fun generateDocURI(did: String, fragment: String): String = "$did#$fragment"
        fun createKeyDocURI(did: String, keyTag: String): String = "$did#$keyTag"
    }

    private fun validateDID(did: String) {
        require(did.isNotEmpty()) { "DID string is required" }
        val parts = did.split(":")
        require(parts.size >= 3 && parts[0] == "did") { "Invalid DID format: '$did'" }
        require(parts[1].isNotEmpty()) { "DID method is required" }
    }

    suspend fun resolve(did: String): DIDDocument {
        validateDID(did)
        return client.resolveDID(did)
    }

    suspend fun create(
        method: String,
        keyType: String? = null,
        network: String? = null,
        services: List<ServiceEndpoint>? = null,
    ): DIDDocument {
        require(method.isNotEmpty()) { "DID method is required" }
        require(SUPPORTED_METHODS.contains(method)) { "Unsupported DID method: '$method'" }
        return client.createDID(method, keyType, network, services)
    }

    suspend fun deactivate(did: String) {
        validateDID(did)
        client.deactivateDID(did)
    }

    suspend fun rotateKey(did: String, keyId: String, keyType: String? = null): DIDDocument {
        validateDID(did)
        require(keyId.isNotEmpty()) { "Key ID is required for rotation" }
        return client.rotateKey(did, keyId, keyType)
    }

    suspend fun resolveWithMetadata(did: String): Triple<DIDDocument, Long, Long> {
        val start = System.currentTimeMillis()
        val document = resolve(did)
        val duration = System.currentTimeMillis() - start
        return Triple(document, duration, System.currentTimeMillis() / 1000)
    }

    suspend fun getVerificationMethods(did: String): List<DIDVerificationMethod> {
        val doc = resolve(did)
        return doc.verificationMethod
    }

    suspend fun getServices(did: String): List<ServiceEndpoint> {
        val doc = resolve(did)
        return doc.service ?: emptyList()
    }

    suspend fun findServiceByType(did: String, serviceType: String): ServiceEndpoint? {
        val services = getServices(did)
        return services.find { it.type == serviceType }
    }

    suspend fun getController(did: String): List<String> {
        val doc = resolve(did)
        return doc.controller ?: listOf(doc.id)
    }

    suspend fun isDeactivated(did: String): Boolean = try {
        resolve(did)
        false
    } catch (e: Exception) {
        e.message?.contains("deactivated") == true
    }

    suspend fun batchResolve(dids: List<String>): Map<String, DIDDocument> {
        return dids.mapNotNull { did ->
            try {
                did to resolve(did)
            } catch (_: Exception) {
                null
            }
        }.toMap()
    }

    fun createDIDDocument(
        id: String,
        controller: List<String>? = null,
        verificationMethods: List<DIDVerificationMethod> = emptyList(),
        services: List<ServiceEndpoint>? = null,
        alsoKnownAs: List<String>? = null,
    ): DIDDocument {
        val now = java.time.Instant.now().toString()
        return DIDDocument(
            context = listOf(
                "https://www.w3.org/ns/did/v1",
                "https://w3id.org/security/suites/ed25519-2020/v1",
            ),
            id = id,
            controller = controller,
            alsoKnownAs = alsoKnownAs,
            verificationMethod = verificationMethods,
            authentication = verificationMethods.map { it.id },
            service = services,
            created = now,
            updated = now,
        )
    }

    fun createVerificationMethod(
        id: String,
        type: String,
        controller: String,
        publicKey: Pair<String?, Map<String, Any>?>,
    ): DIDVerificationMethod {
        return DIDVerificationMethod(
            id = id,
            type = type,
            controller = controller,
            publicKeyMultibase = publicKey.first,
            publicKeyJwk = publicKey.second,
        )
    }

    fun addVerificationMethod(
        doc: DIDDocument,
        method: DIDVerificationMethod,
    ): DIDDocument {
        return doc.copy(
            verificationMethod = doc.verificationMethod + method,
            authentication = doc.authentication + method.id,
        )
    }

    fun removeVerificationMethod(
        doc: DIDDocument,
        methodId: String,
    ): DIDDocument {
        return doc.copy(
            verificationMethod = doc.verificationMethod.filter { it.id != methodId },
            authentication = doc.authentication.filter { it != methodId },
            assertionMethod = doc.assertionMethod?.filter { it != methodId },
            keyAgreement = doc.keyAgreement?.filter { it != methodId },
        )
    }

    fun addService(doc: DIDDocument, service: ServiceEndpoint): DIDDocument {
        return doc.copy(
            service = (doc.service ?: emptyList()) + service,
        )
    }

    fun removeService(doc: DIDDocument, serviceId: String): DIDDocument {
        return doc.copy(
            service = doc.service?.filter { it.id != serviceId },
        )
    }
}

File: sdk/swift/Sources/KYCVault/DIDResolver.swift
import Foundation

public class DIDResolver {
    private let client: KYCVaultClient

    public init(client: KYCVaultClient) {
        self.client = client
    }

    private func validateDID(_ did: String) throws {
        guard !did.isEmpty else {
            throw DIDResolverError.validation("DID string is required")
        }
        let parts = did.split(separator: ":")
        guard parts.count >= 3, parts[0] == "did" else {
            throw DIDResolverError.validation("Invalid DID format: '\(did)'")
        }
        guard !parts[1].isEmpty else {
            throw DIDResolverError.validation("DID method is required")
        }
    }

    public func resolve(_ did: String) async throws -> DIDDocument {
        try validateDID(did)
        return try await client.resolveDID(did)
    }

    public func create(
        method: String,
        keyType: String? = nil,
        network: String? = nil,
        services: [ServiceEndpoint]? = nil
    ) async throws -> DIDDocument {
        guard !method.isEmpty else {
            throw DIDResolverError.validation("DID method is required")
        }
        let supported = ["key", "web", "ethr", "ion", "indy"]
        guard supported.contains(method) else {
            throw DIDResolverError.unsupportedMethod(method)
        }
        return try await client.createDID(method: method, keyType: keyType, network: network, services: services)
    }

    public func deactivate(_ did: String) async throws {
        try validateDID(did)
        try await client.deactivateDID(did)
    }

    public func rotateKey(did: String, keyId: String, keyType: String? = nil) async throws -> DIDDocument {
        try validateDID(did)
        guard !keyId.isEmpty else {
            throw DIDResolverError.validation("Key ID is required for rotation")
        }
        return try await client.rotateKey(did: did, keyId: keyId, keyType: keyType)
    }

    public func resolveWithMetadata(_ did: String) async throws -> (document: DIDDocument, duration: TimeInterval, resolvedAt: Int) {
        let start = Date()
        let doc = try await resolve(did)
        let duration = Date().timeIntervalSince(start)
        return (doc, duration, Int(Date().timeIntervalSince1970))
    }

    public func getVerificationMethods(_ did: String) async throws -> [DIDVerificationMethod] {
        let doc = try await resolve(did)
        return doc.verificationMethod
    }

    public func getServices(_ did: String) async throws -> [ServiceEndpoint] {
        let doc = try await resolve(did)
        return doc.service ?? []
    }

    public func findServiceByType(_ did: String, serviceType: String) async throws -> ServiceEndpoint? {
        let services = try await getServices(did)
        return services.first { $0.type == serviceType }
    }

    public func getController(_ did: String) async throws -> [String] {
        let doc = try await resolve(did)
        return doc.controller ?? [doc.id]
    }

    public func isDeactivated(_ did: String) async throws -> Bool {
        do {
            let _ = try await resolve(did)
            return false
        } catch let error as DIDResolverError where error.localizedDescription.contains("deactivated") {
            return true
        }
    }

    public func batchResolve(_ dids: [String]) async throws -> [String: DIDDocument] {
        var results: [String: DIDDocument] = [:]
        try await withThrowingTaskGroup(of: (String, DIDDocument).self) { group in
            for did in dids {
                group.addTask {
                    let doc = try await self.resolve(did)
                    return (did, doc)
                }
            }
            for try await (did, doc) in group {
                results[did] = doc
            }
        }
        return results
    }

    public static func parseDID(_ did: String) throws -> (method: String, methodSpecificId: String) {
        let parts = did.split(separator: ":")
        guard parts.count >= 3, parts[0] == "did" else {
            throw DIDResolverError.invalidDID(did)
        }
        return (String(parts[1]), parts.dropFirst(2).joined(separator: ":"))
    }

    public static func isValidDID(_ did: String) -> Bool {
        (try? parseDID(did)) != nil
    }

    public static func generateDocURI(did: String, fragment: String) -> String {
        "\(did)#\(fragment)"
    }

    public static func createKeyDocURI(did: String, keyTag: String) -> String {
        "\(did)#\(keyTag)"
    }

    public func createDIDDocument(
        id: String,
        controller: [String]? = nil,
        verificationMethods: [DIDVerificationMethod] = [],
        services: [ServiceEndpoint]? = nil,
        alsoKnownAs: [String]? = nil
    ) -> DIDDocument {
        DIDDocument(
            context: [
                "https://www.w3.org/ns/did/v1",
                "https://w3id.org/security/suites/ed25519-2020/v1",
            ],
            id: id,
            controller: controller,
            alsoKnownAs: alsoKnownAs,
            verificationMethod: verificationMethods,
            authentication: verificationMethods.map { $0.id },
            assertionMethod: nil,
            keyAgreement: nil,
            capabilityInvocation: nil,
            capabilityDelegation: nil,
            service: services,
            created: ISO8601DateFormatter().string(from: Date()),
            updated: ISO8601DateFormatter().string(from: Date())
        )
    }

    public func createVerificationMethod(
        id: String,
        type: String,
        controller: String,
        publicKey: (multibase: String?, jwk: [String: AnyCodable]?)
    ) -> DIDVerificationMethod {
        DIDVerificationMethod(
            id: id,
            type: type,
            controller: controller,
            publicKeyMultibase: publicKey.multibase,
            publicKeyJwk: publicKey.jwk,
            blockchainAccountId: nil
        )
    }
}

public enum DIDResolverError: Error, LocalizedError {
    case validation(String)
    case unsupportedMethod(String)
    case invalidDID(String)

    public var errorDescription: String? {
        switch self {
        case .validation(let msg): return msg
        case .unsupportedMethod(let method): return "Unsupported DID method: '\(method)'"
        case .invalidDID(let did): return "Invalid DID: \(did)"
        }
    }
}
